sdk/helper/aliasmetadata/alias_metadata.go

```diff
@@ -0,0 +1,195 @@
+package aliasmetadata
+
+/*
+	aliasmetadata is a package offering convenience and
+	standardization when supporting an `alias_metadata`
+	field in a plugin's configuration. This then controls
+	what alias metadata is added to an Auth during login.
+
+	To see an example of how to add and use it, check out
+	how these structs and fields are used in the AWS auth
+	method.
+
+	Or, check out its acceptance test in this package to
+	see its integration points.
+*/
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/strutil"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+// Fields is for configuring a back-end's available
+// default and additional fields. These are used for
+// providing a verbose field description, and for parsing
+// user input.
+type Fields struct {
+	// The field name as it'll be reflected in the user-facing
+	// schema.
+	FieldName string
+
+	// Default is a list of the default fields that should
+	// be included if a user sends "default" in their list
+	// of desired fields. These fields should all have a
+	// low rate of change because each change can incur a
+	// write to storage.
+	Default []string
+
+	// AvailableToAdd is a list of fields not included by
+	// default, that the user may include.
+	AvailableToAdd []string
+}
+
+func (f *Fields) all() []string {
+	return append(f.Default, f.AvailableToAdd...)
+}
+
+// FieldSchema takes the default and additionally available
+// fields, and uses them to generate a verbose description
+// regarding how to use the "alias_metadata" field.
+func FieldSchema(fields *Fields) *framework.FieldSchema {
+	return &framework.FieldSchema{
+		Type:        framework.TypeCommaStringSlice,
+		Description: description(fields),
+		DisplayAttrs: &framework.DisplayAttributes{
+			Name:  fields.FieldName,
+			Value: "default,field1,field2",
+		},
+		Default: []string{"default"},
+	}
+}
+
+// NewHandler instantiates a Handler to be embedded in your config.
+func NewHandler(fields *Fields) Handler {
+	return &handler{
+		fields: fields,
+	}
+}
+
+// Handler is an interface for the helper methods you get on your
+// config when you embed the Handler.
+type Handler interface {
```

Comment:
Curious if we actually need an interface for this or if we can just use a concrete type?

Reply:
I used it for implementation hiding as discussed [here](https://www.informit.com/articles/article.aspx?p=1760496&seqNum=4). The main reason was, I don't want folks to directly edit the `AliasMetadata` field because I don't want them to directly put `default` in and cause a bug. Normally I would just make it private, but if I did, the field wouldn't be properly converted to JSON for storage. So, I used an interface instead to make it un-editable.

Reply:
I see, it makes sense that we don't want to allow the values to be set manually. One option would be to leave them as private variables but add `UnmarshalJSON` and `MarshalJSON` functions so we can still output the field in json. Having an interface implies the underlying implementation can be swapped out and I'm not sure we want to encourage that since the implementation is tied to the underlying storage entries for a given backend.

Reply:
Huh, yes, I like that approach. Will flip over to it.

Reply:
https://github.com/hashicorp/vault/pull/8783/commits/f180042327b68ba3c6971e098478317348902aef